### PR TITLE
Add Open Trace option to Menu & Trace Explorer

### LINKED ISSUE #159: Add “Open Trace” option to File menu
Adding an “Open Trace” option in the File menu keeps this operation intuitive (by imitating how any file is opened).

## Position in menu
**Suggestion:** “Open Trace” could be in a separate menu section to distinguish it from the other standard IDE open operations:
file-menu-open-trace

**Alternative:** “Open Trace” could also be further down to avoid breaking up the more standard IDE operations. This would create a bit more discontinuity in the menu list since all the “Open” and the “Save” operations would not be grouped:
file-menu-open-trace-2

### LINKED ISSUE #160: Add “Open Trace” button to trace extension sidebar
Adding an “Open Trace” button to trace extension sidebar keeps this operation intuitive by imitating how a folder is opened in VS Code.

* This button and text is only visible when no traces are open.
* If all currently open traces are closed, this text and button reappears.

open-trace-button2

Text and button style:
vscode-open-folder-button

### PR DESCRIPTION
**Pull request by muddana-satish**
fixes #160

Signed-off-by: muddana-satish <satish.muddana@ericsson.com>

### CODE REVIEW

**Comment by muddana-satish** [comment]
Adding few screenshots:

Added a new menu item under File menu.

image

I have added it right under `Open`. Thoughts? 

Added a new button in Trace Explorer.
image

Using 'Open Trace' option just opens the trace. It doesn't load it to the file explorer. (similar to Open functionality)

On Click of 'Open Trace' button in Trace Explorer:
Kapture 2020-11-24 at 00 27 31

On Click of 'Open Trace' in File menu.
Kapture 2020-11-24 at 00 32 49

File: viewer-prototype/src/browser/trace-viewer/trace-viewer-frontend-module.ts [hunk 1]
```diff
@@ -4,7 +4,8 @@ import { TraceViewerWidget, TraceViewerWidgetOptions } from './trace-viewer';
 import { TraceViewerContribution } from './trace-viewer-contribution';
 import { TraceViewerEnvironment } from '../../common/trace-viewer-environment';
 import { TraceServerUrlProvider } from '../../common/trace-server-url-provider';
-import { CommandContribution } from '@theia/core/lib/common';
+import { CommandContribution, MenuContribution } from '@theia/core/lib/common';
```
Comment (bhufmann) [inline viewer-prototype/src/browser/trace-viewer/trace-viewer-frontend-module.ts:7]: I wonder if we should add the command to the main File menu (see https://github.com/theia-ide/theia-trace-extension/issues/159#issuecomment-733749083). So, I think the feature to add it to the main menu should be a different PR. That would allow us to merge the feature of "Open Trace" button (#160) first and decide the way forward for #159.

**Review by bhufmann — CHANGES_REQUESTED** [review]
I really like the open trace button in the trace explorer and the command short-cut with CTRL+SHIFT+P. Please have a separate PR for the menu contribution as described in my comments.

**Comment by muddana-satish** [comment]
> I really like the open trace button in the trace explorer and the command short-cut with CTRL+SHIFT+P. Please have a separate PR for the menu contribution as described in my comments.

Thanks @bhufmann  for the update.
I removed the menu contribution and added just the 'open trace' button in trace explorer as part of this PR. For the shortcut, I will create a new PR and work on it.

File: viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import { TraceViewerWidget, TraceViewerWidgetOptions } from './trace-viewer';
+import { WidgetOpenHandler } from '@theia/core/lib/browser';
+import { Command, CommandRegistry, CommandContribution } from '@theia/core';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+import { FileDialogService, OpenFileDialogProps } from '@theia/filesystem/lib/browser';
+
+export namespace TraceMenuCommand {
+    export const OPEN: Command = {
+        id: 'open-trace',
+        label: 'Open Trace'
```
[the remark this answers is not in the document]
Comment (ebugden) [reply viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts]: I suggest leaving it singular as "Open Trace". It's expected that in a command like this you can explicitly decide to open multiple traces in the same folder. For example, in the VS Code "Open File" dialogue box you can select multiple files to open (when holding Ctrl).

Having the implicit ability to open multiple traces by selecting a directory (but only for a certain trace type) isn't something I would necessarily want to advertise from a UX perspective. It shows inconsistencies in the interaction rules of the tool universe and could create distrust in a user.

Are there other actions that are implicit in this "Open" operation? Is it possible to open a Trace Compass Experiment with this?

**Review by ebugden — CHANGES_REQUESTED** [review]
Apart from a couple details the visuals look good on my end! I went for maximum visual pickiness and I have no idea how difficult it is to implement these while keeping responsiveness, prioritizing elegant code, etc. So let me know. Especially for MVP I'm willing to loosen criteria to avoid code hacks. 

Disclaimer for other reviewers: I just reviewed the visual style. I didn't run the code or check if it breaks anything. 

Thanks for working on this! It's my first time reading code from this project and I understood your changes and what was going on so that says something about your code clarity 🍰

File: viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhufmann) [reply viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts]: > I suggest leaving it singular as "Open Trace". It's expected that in a command like this you can explicitly decide to open multiple traces in the same folder. For example, in the VS Code "Open File" dialogue box you can select multiple files to open (when holding Ctrl). The "Open Folder" option also allows you to open multiple folders.
> 
Just a note, that this will require that the user knows the details of the used tracer and which traces the user needs to select and where these traces are located. In my experience, this is often not known. 
For example, for an LTTng session there is a non-tivial directory structure with kernel in the session root directory and one or more UST traces further down in the tree. Don't forget that CTF traces are directories.
The implementatoin supports opening LTTng kernel and UST together for trace correlation by selecting the session root (which is main use case of Trace Compass). 

In the future, I think will have to provide a service ("trace detection and experiment configuration service") for the user to make it easy to open a set of traces of a system that needs to be analyzed and put together. This service will have to be configurable for different tracers. It's related to #8. 

> Having the implicit ability to open multiple traces by selecting a directory (but only for a certain trace type) isn't something I would necessarily want to advertise from a UX perspective. It shows inconsistencies in the interaction rules of the tool universe and could create distrust in a user.
> 
I understand and it make sense. We don't want to break user experience, while giving the user a good experience for trace viewer.

> Are there other actions that are implicit in this "Open" operation? Is it possible to open a Trace Compass Experiment with this?
No, there is no other implicit "Open" operation. The only way to open an experiment with multiple traces is to select a directory with multiple CTF in it.

Now, we need to think about the way forward and what to include in the MVP. Opening multiple traces together is required for the MVP. I see the following 2 options (should be discussed in a different issue tracker):
Option 1)
Keep the implementation as is. Name all commands "Open Trace". Allow to select a directory and scan the sub-directories for CTF traces and open them together. 

Advantages: is that it's already supported. 
Disadvantages: It has an implicit and unexpected behavior when selecting a directory. Selecting a large directory can cause some troubles too.

Option 2)
-  From the context sensitive-menu of file explorer: Support opening a single trace only.
-  From new open button or CTRL+SHIFT+P: Allow user to select multiple directories and pass them to the open trace command. In the open trace command we post each trace to the trace server and then post the experiment containing the traces which will be shown in the trace explorer.

Advantages: Explicit opening of multiple traces in a single experiment. 
Disadvantages: More involving from implementation point of view (might not be ready for the MVP). User has to know about which traces to select.

**Comment by bhufmann** [comment]
@ebugden I opened #189 to track the discussion of the open trace behavior for multiple traces (for MVP) in a separate tracker.

@muddana-satish, this PR is independent to the outcome of that discussion. I suggest to add the "Open Trace" button in the trace explorer and CTRL+SHIT+P behavior as you proposed (without the file menu option) and merge it. Later on we can update the implementation based on the decision.

File: viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx [hunk 3]
```diff
@@ -112,66 +116,72 @@ export class TraceExplorerWidget extends ReactWidget {
...
+                    </div>
+                    <div className='trace-explorer-panel-content'>
+                        {this.renderTooltip()}
+                    </div>
                 </div>
+            </div>;
+        }
+
+        return <div className='theia-navigator-container' tabIndex={0}>
+            <div className='center'>{'You have not yet opened a Trace.'}</div>
```
[the remark this answers is not in the document]
Comment (muddana-satish) [reply viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx:181]: @ebugden  I'm making use of the existing theia styles. I have not created any particular style for 'Open Trace' button. In  other words, the newly added button is in consistent with the other buttons in theia. 

To keep it in consistent with VSCode, I think we would need to update the theia-styles (not trace extension styles).

File: viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx [hunk 4]
```diff
@@ -112,66 +116,72 @@ export class TraceExplorerWidget extends ReactWidget {
...
+            <div className='open-workspace-button-container'>
+                <button className='theia-button open-workspace-button' title='Select a Trace to open'
```
[the remark this answers is not in the document]
Comment (muddana-satish) [reply viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx]: > @muddana-satish, this PR is independent to the outcome of that discussion. I suggest to add the "Open Trace" button in the trace explorer and CTRL+SHIT+P behavior as you proposed (without the file menu option) and merge it. Later on we can update the implementation based on the decision.

I have updated the code. Added Command Palette behaviour. Please review the changes.

File: viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx [hunk 5]
```diff
@@ -112,66 +116,72 @@ export class TraceExplorerWidget extends ReactWidget {
...
+                    onClick={this.handleOpenTrace}>{'Open Trace(s)'}</button>
```
[the remark this answers is not in the document]
Comment (muddana-satish) [reply viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx]: > Apart from a couple details the visuals look good on my end! I went for maximum visual pickiness and I have no idea how difficult it is to implement these while keeping responsiveness, prioritizing elegant code, etc. So let me know. Especially for MVP I'm willing to loosen criteria to avoid code hacks.
> 
> Disclaimer for other reviewers: I just reviewed the visual style. I didn't run the code or check if it breaks anything.
> 
> Thanks for working on this! It's my first time reading code from this project and I understood your changes and what was going on so that says something positive about your code clarity 🍰

Thanks for reviewing the changes. I have updated few UI changes as per your suggestions.

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks for the contribution.

**Comment by mirsky-work** [comment]
@muddana-satish, great UX improvement, thank you!

I know I'm late to the party, but wouldn't be better to change the string "_You have not yet opened a trace._" to something like "_No open traces_"? It's shorter and makes more sense if we consider the flow when a user opens a trace and then closes it. Of course we can defer this minor change to another PR.

**Comment by muddana-satish** [comment]
> @muddana-satish, great UX improvement, thank you!
> 
> I know I'm late to the party, but wouldn't be better to change the string "_You have not yet opened a trace._" to something like "_No open traces_"? It's shorter and makes more sense if we consider the flow when a user opens a trace and then closes it. Of course we can defer this minor change to another PR.

I think the intention is to keep it similar to what VSCode shows (screenshots in #160). I do not have a strong opinion on this. I am okay with the suggest change if it's okay for others as well.

File: viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebugden) [reply viewer-prototype/src/browser/trace-viewer/trace-viewer-menu.ts]: > This will require that the user knows the details of the used tracer and which traces the user needs to select and where these traces are located. In my experience, this is often not known.

For the MVP I'm still applying the assumption that users have the perfect theoretical background for using the tool. I agree that this solution for opening data sources does not represent the complex use cases and guidance required in the final tool. 

> Now, we need to think about the way forward and what to include in the MVP. Opening multiple traces together is required for the MVP.

I'll respond directly in the new issue #189.

**Comment by ebugden** [comment]
> @muddana-satish I think the intention is to keep it similar to what VSCode shows

Yes that was the intention. If the extension uses phrasing that is consistent with the rest of VS Code it will feel well integrated to users and so they intuitively trust it more easily. For most users, this sentence is probably their first interaction with the extension and we're wired to interpret anomalies as potentially "off". 

If Theia uses different phrasing and/or 95% of users are using it with Theia, then integration with VS Code can be less of a factor. 

---

> @mirsky-work wouldn't be better to change the string "You have not yet opened a trace." to something like "No open traces"? It's shorter and makes more sense if we consider the flow when a user opens a trace and then closes it.

I'm not sure what you mean by "No open traces" makes more sense considering the user flow. This button is only visible if no traces are open (initially forgot to specify, just updated issue #160). If a user closes all their open traces and they read the text again, I think it's still clear that the text refers to the future. 

If I were to hypothesize as to why VS Code uses longer wording I would guess:
* **It's human:** It feels gentler and more human to read a full sentence. I like this because it's an easy way to make the tool feel approachable (and less cold and utilitarian). 
* **It's specific:** In "**You have not yet \<performed an action\>**", "**You**" specifies that it's the user's responsibility to perform the action and "**yet**" specifies that they are able to perform the action. Ex. "No trace server" vs. "You have not yet started the trace server". Granted it's debatable whether this extra clarity is required here.  

---

**tl;dr** "You have not yet opened a trace." is specific, approachable, and consistent with VS Code. But if the full sentence is inconsistent with Theia or you guys much prefer "No open traces" you can go for it.

File: viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx [hunk 6]
```diff
@@ -112,66 +116,72 @@ export class TraceExplorerWidget extends ReactWidget {
...
+                    </div>
                 </div>
+            </div>;
+        }
+
+        return <div className='theia-navigator-container' tabIndex={0}>
+            <div className='center'>{'You have not yet opened a trace.'}</div>
+            <div className='open-workspace-button-container'>
+                <button className='theia-button open-workspace-button' title='Select a trace to open'
+                    onClick={this.handleOpenTrace}>{'Open Trace'}</button>
```
Comment (ebugden) [inline viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx:184]: Could you post an updated screenshot of this?

**Review by ebugden — CHANGES_REQUESTED** [review]
As a last check, could you post an updated screenshot? Thanks for the updates!

**Comment by mirsky-work** [comment]
@ebugden, I was not aware that the phrasing complies with existing phrasing in VS Code, in that case I agree we should strive to comply.

> It's specific: In "You have not yet <performed an action>", "You" specifies that it's the user's responsibility to perform the action and "yet" specifies that they are able to perform the action. Ex. "No trace server" vs. "You have not yet started the trace server". Granted it's debatable whether this extra clarity is required here.

What about a future scenario, in which the user connects to a remote server, where there might be traces opened by other users?

**Comment by bhufmann** [comment]
Any objections me merging this patch? I think it's a great addition.

**Comment by mirsky-work** [comment]
> Any objections me merging this patch? I think it's a great addition.

LGTM!

File: viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx (hunk 6 above)
Comment (muddana-satish) [reply viewer-prototype/src/browser/trace-explorer/trace-explorer-widget.tsx:184]: Sure.

**Comment by muddana-satish** [comment]
> As a last check, could you post an updated screenshot? Thanks for the updates!

@ebugden 

Trace Explorer:
image

Command Palette:
image

**Comment by ebugden** [comment]
> Button description text "You have not yet opened a Trace." should be left aligned

"You have not yet opened a trace." is still centered here.

> I'm making use of the existing theia styles. I have not created any particular style for 'Open Trace' button. In other words, the newly added button is in consistent with the other buttons in theia.

If you mean that the the text above similar Theia side panel buttons is typically centered then let's leave it centered. Otherwise, I suggest left-aligning "You have not yet opened a trace." to match VS Code.

I think centered text actually looks prettier in this particular case but:
* Longer text blocks will need to be left-aligned for readability (ex. "Before you can view a trace, the trace server must..." #191 [(comment)](#191 (comment))) and left-aligning all text blocks makes text display consistent
* Left-aligned text stays in the same place when the side panel changes width

# Centered (current version)
image

# Left-aligned (to match VS Code)
image

**Comment by muddana-satish** [comment]
> If you mean that the the text above similar Theia side panel buttons is typically centered then let's leave it centered. Otherwise, I suggest left-aligning "You have not yet opened a trace." to match VS Code.

@ebugden Yes. In theia, it is centered. So I believe it's best to leave it centered to keep it consistent. I do not know how CSS theia classes take effect when we use it in VSCode. We may definitely need to revisit if it doesn't align correctly in VSCode.

**Comment by bhufmann** [comment]
I'm going to merge it now. We can revisit the style (centered or not) later on when migrating to VSCode. I like it centered as well :-)

**Comment by ebugden** [comment]
> Yes. In theia, it is centered. So I believe it's best to leave it centered to keep it consistent.

Sounds good! Thank you for answering all my detailed questions!
